provider/oracle/storage_volumes.go

```diff
@@ -0,0 +1,514 @@
+		return false, errors.Trace(err)
+	}
+	return string(details.State) == desiredStatus, nil
+}
+
+// waitForMachineStatus will ping the volume until the timeout duration is reached or an error appeared
+func (o *oracleVolumeSource) waitForResourceStatus(
+	fetch func(name string, desiredStatus string) (complete bool, err error),
+	name, state string, timeout time.Duration) error {
+
+	// chan user for errors
+	errChan := make(chan error)
+	// chan used for timeout
+	done := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				done, err := fetch(name, state)
+				if err != nil {
+					errChan <- err
+					return
+				}
+				if done {
+					errChan <- nil
+					return
+				}
+				time.Sleep(2 * time.Second)
+			}
+		}
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-time.After(timeout):
+		done <- true
+		return errors.Errorf(
+			"timed out waiting for resource %q to transition to %v",
+			name, state,
+		)
+	}
+
+	return nil
+}
+
+// ListVolumes lists the provider volume IDs for every volume
+// created by this volume source.
+func (s *oracleVolumeSource) ListVolumes() ([]string, error) {
+	tag := fmt.Sprintf("%s=%s", tags.JujuModel, s.modelUUID)
+	filter := []oci.Filter{
+		oci.Filter{
+			Arg:   "tags",
+			Value: tag,
+		},
+	}
+	volumes, err := s.api.AllStorageVolumes(filter)
+	if err != nil {
+		return nil, errors.Annotate(err, "listing volumes")
+	}
+
+	ids := make([]string, 0, len(volumes.Result))
+	for _, volume := range volumes.Result {
+		ids = append(ids, volume.Name)
+	}
+
+	return ids, nil
+}
+
+// DescribeVolumes returns the properties of the volumes with the
+// specified provider volume IDs.
+func (s *oracleVolumeSource) DescribeVolumes(volIds []string) ([]storage.DescribeVolumesResult, error) {
+	n := len(volIds)
+	if volIds == nil || n == 0 {
+		return []storage.DescribeVolumesResult{}, nil
+	}
+
+	tag := fmt.Sprintf("%s=%s", tags.JujuModel, s.modelUUID)
+	filter := []oci.Filter{
+		oci.Filter{
+			Arg:   "tags",
+			Value: tag,
+		},
+	}
+
+	result := make([]storage.DescribeVolumesResult, 0, n)
+	volumes, err := s.api.AllStorageVolumes(filter)
+	if err != nil {
+		return nil, errors.Annotatef(err, "descrie volumes")
+	}
+	asMap := map[string]ociResponse.StorageVolume{}
+	for _, val := range volumes.Result {
+		asMap[val.Name] = val
+	}
+	for i, volume := range volIds {
+		if vol, ok := asMap[volume]; ok {
+			volumeInfo := &storage.VolumeInfo{
+				VolumeId:   vol.Name,
+				Size:       uint64(vol.Size) / 1024 / 1024 / 1024,
+				Persistent: true,
+			}
+			v := storage.DescribeVolumesResult{
+				VolumeInfo: volumeInfo,
+			}
+			result[i] = v
+		} else {
+			result[i].Error = errors.NotFoundf("%s", volume)
+		}
+
+	}
+
+	return result, nil
+}
+
+// DestroyVolumes destroys the volumes with the specified provider
+// volume IDs.
+func (s *oracleVolumeSource) DestroyVolumes(volIds []string) ([]error, error) {
+	results := make([]error, len(volIds))
```

Comment:
if no volumes are specified, this function becomes a noop, and returns ```nil,nil```. If you prefer, I can add a check at the beginning that returns right away, but the end result is the same.